### ISSUE TITLE
promtool: backfill: very slow when operating across long durations

### ISSUE THREAD

**Issue by stevekuznetsov** [prometheus/prometheus#8915]
I am generating data across the entire retention period to backfill a testing setup for Prometheus. I use 100 unique metric label sets, 125 samples per metric, randomly distributed across the entire retention period. Actually generating and encoding this data into the OpenMetrics format takes on the order of microseconds. Running `promtool tsdb create-blocks-from openmetrics` takes 30s-1m. This seems pretty excessive for runtime. In my testing, it seems like this is scaling with the number of unique blocks that need to be inserted. I'm wondering if a maintainer or author of the tool could weigh in on why the backfill would be so slow and what approaches could be taken to improve it. I'm happy to work on this, but after reading the backfill code it looks like the issue might be in the way in which the tool is interacting with the `tsdb`? Could separate blocks be written in parallel? Is there a way to change the configuration to allow for quicker insertions when the user has no previous data (in a test instance?). Is the backfill code using the same write-paths as the production system does on scrapes?

Would love some pointers or help :)

cc @tomwilkie  @roidelapluie @asquare14

**Comment by stevekuznetsov** [comment]
I've tried aligning the generated timestamps on the data to scrape boundaries, FWIW - this did not make any difference. Generating all of the data within one block (the most recent one) made the backfill very quick.

**Comment by roidelapluie** [comment]
Yes, we need a flag to set the block duration. There is already an issue
for this.

Le jeu. 10 juin 2021 à 18:03, Steve Kuznetsov ***@***.***> a
écrit :

**Comment by stevekuznetsov** [comment]
Sounds like that's https://github.com/prometheus/prometheus/issues/8388 - I'll take a stab at this :)

**Comment by roidelapluie** [comment]
Closing as duplicate of #8388. Another pull request #8197 has been merged related to this as well.
